### ISSUE TITLE
Pen Size 0 for bresenham-like lines (feature request)

### ISSUE THREAD

**Issue by Tegmen** [LLK/scratch-flash#1174]
Currently there is no way to set pixels exactly. Pen size 1 draws an anti-aliased circle (resp. line), which results in pixels being drawn with an opacity <100%. One can use pen size 2 to draw a pixel with full opacity, however this affects neighbour pixels as well.
My proposition: Allow the pen size being set to 0, which would change the mode of lines/dots being drawn from AS3's lineTo() to a bresenham-like line without anti-aliasing ([EFLA by Po-Han Lin](http://www.simppa.fi/blog/extremely-fast-line-algorithm-as3-optimized/) is probably the fastest way to do so). This would not only make it possible to set pixels exactly, but also increase the speed at which lines are being drawn.
I am aware of AS3's graphics object not having a setPixel() function, so to implement a bresenham-like algorithm, you had to change a lot of the current code. So maybe this is more a feature request for Scratch 3.0.

**Comment by Tegmen** [comment]
primitives/PenAndMotionPrims.as line 287:
`if (s != null) s.setPenSize(Math.max(1, Math.min(960, Math.round(interp.numarg(b, 0)))));`

Why does it set upper and lower bounds, when Sprite already does that?

scratch/ScratchSprite.as setPenSize() line 264 :
`penWidth = Math.max(1, Math.min(Math.round(n), 255));`

**Comment by TheLogFather** [comment]
To draw single pixels, try offsetting the x & y position by 0.5...

**Comment by Tegmen** [comment]
I came to similar results when testing for my pen Tetris (my tests show that (0.4/0.6) gives the best results). This DOES only affect a single pixel, but the opacity is how i mentioned less than 100%. 
Try to colour a pixel black, then force a canvas update or wait a frame and paint a white pixel over it. The result wont be a white pixel, but a grey one. The closer the x/y to 0.4/0.6, the higher the opacity, but it is never 100% and usually needs about 2 or 3 pen-up-pen-downs to colour a pixel in the intended colour.
But its not just the accuracy, a bresenham-like line would also be a lot faster than AS3's lineTo(), which is a complex floating point algorithm. But I think that the interpretation of the Scratch blocks is the bottleneck here, so the total speed improvements might be minor.
@MartinBraendli

**Comment by TheLogFather** [comment]
Yes, it's not 100% opacity, so multiple pen-downs needed. Interesting about the 0.6 – hadn't tried that.
Never saw the point of attempting to emulate Bresenham-like drawing in Scratch (except for educational purposes), since all its vars are floats anyway (not to mention how slow block interpretation is...)

**Comment by griffpatch** [comment]
To draw a pixel I usually do this:

set pen size to 1
goto x+0.5, y+0.4
pen down
change y by 0.6

I think that's right...  You can flip it around too and doa change x if you
prefer... (just reverse the 0.5 and 0.4 too)

@griffpatch

**Comment by TheLogFather** [comment]
Wow, that actually works for overdrawing too!
(i.e. draw a black pixel, wait a frame, draw a white pixel –> nothing left of black.)
:)

**Comment by griffpatch** [comment]
Yeah, I discovered it while trying to create that fast text renderer... :)
I'm sure it will break horribly when we move to scratch 3 ;)

**Comment by Tegmen** [comment]
Nice hack! Now i feel kind of stupid for not having tried out myself moving the pen inside the same pixel. This is better than "pen-stamping" multiple times, since it respects the alpha value of the pen and might also be a bit faster.

**Comment by Tegmen** [comment]
After some testing: When using opacity, griffpatch's hack uses an opacity a bit too high. A Fully black pixel with an opacity of 50% on a white background should give a pixel with colour value 0x949494 (Photoshop and pen size 10 both give those results). The hack however gives a pixel with colour 0x8d8d8d, which is too dark. 
I really like this hack, but it is not perfect (slow speed, opacity error). Therefore I still hope, that Scratch 3.0 will have a better method, for setting single pixels. Also, in HTML5 the speed difference between bresenham-like lines and AS3-emulated lines will be noticeable.
